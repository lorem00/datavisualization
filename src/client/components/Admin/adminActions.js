import axios from 'axios'
import rest from '../../components/common/utils/restconfig'

export const GET_USERS_PENDING = 'GET_USERS_PENDING';
export const GET_USERS_SUCCESS = 'GET_USERS_SUCCESS';
export const GET_USERS_ERROR = 'GET_USERS_ERROR';
export const GET_USERS = 'GET_USERS';

export function getAllUsers() {
    return dispatch => {
        dispatch(setAdminPending(true));
        dispatch(setAdminSuccess(false));
        dispatch(setAdminError(false));

        getAllUsersApi(function(error, list) {
            dispatch(setAdminPending(false));
            if (!error) {
                dispatch(setAdminSuccess(true));
                dispatch(getAllUserss(list));
            } else {
                dispatch(setAdminError(true));
            }
        });
    }
}

function getAllUsersApi( callback ) {
    let options = {
        headers: Object.assign({'Content-Type': 'application/json'})
    };
    axios.get(rest.GET_ALL_USERS, options)
        .then((response) => {
            if(response.data!=null){

                return callback(null, response.data);
            }else{
                return callback(new Error('Could not get All Users'));
            }
        }).catch((err) => {
            return callback(new Error('Could not get All Users'));
        });
}

function setAdminPending(isPending) {
    return {
        type: GET_USERS_PENDING,
        isPending
    };
}

function setAdminSuccess(isSuccess) {
    return {
        type: GET_USERS_SUCCESS,
        isSuccess
    };
}

function setAdminError(isError) {
    return {
        type: GET_USERS_ERROR,
        isError
    }
}

function getAllUserss(users){
    return {
        type: GET_USERS,
        users
    }
}