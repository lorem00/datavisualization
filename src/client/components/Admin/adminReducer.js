/**
 * Created by ali_jalbani on 10/18/17.
 */
import {GET_USERS_PENDING,
    GET_USERS_SUCCESS,
    GET_USERS_ERROR,
    GET_USERS} from './adminActions'

const INITIAL_STATE = {
    isSuccess: false,
    isPending: false,
    isError: false,
    users: [],
    token: '',
    message:''
};
export default function adminReducer(state = INITIAL_STATE, action) {

    switch (action.type) {
        case GET_USERS_PENDING:
            return Object.assign({}, state, {
                isPending: action.isPending
            });

        case GET_USERS_SUCCESS:
            return Object.assign({}, state, {
                isSuccess: action.isSuccess
            });

        case GET_USERS_ERROR:
            return Object.assign({}, state, {
                isError: action.isError,
                message:'Could not get users.'
            });

        case GET_USERS:
            // alert(JSON.stringify(action));
            return Object.assign({}, state, {
                users: action.users
            });

        default:
            return state;
    }
}