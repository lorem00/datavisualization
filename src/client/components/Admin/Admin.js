import React, { Component } from 'react';
import { connect } from 'react-redux';
import { BootstrapTable, TableHeaderColumn } from 'react-bootstrap-table';
import Constants from './../common/utils/Constants';
import Authorization from '../../components/common/utils/Authorization';
import { getAllUsers } from './adminActions';

export class AdminComponent extends Component {
    constructor(props) {
        super(props);
        this.state = {
            users: []
        };
    }

    componentWillMount() {
        //did this for unit tests to pass
        if (this.props.getAllUsers) {
            this.props.getAllUsers();
        }
    }

    editUser(id, field, value) {
        // alert(id + " " + field + " " + value);
        let actualvalue = '';
        let actualfield = '';
        switch(field){
            case 'roledisplay':
                actualfield = 'role';
                actualvalue = Constants.ROLES.find(role => role.display == value).id;
                break;
            default:
                actualfield = field;
                actualvalue = value;
        }
        //this.props.dispatch(saveUserField(id, actualfield, actualvalue));
    }

    format(users){
        for(let k in users){
            if(users[k].roles !=undefined) {
                let rolesarr = [];
                for(let l in users[k].roles){
                    rolesarr.push(Constants.ROLES.filter((obj)=>obj.name==users[k].roles[l])[0].display);
                }
                users[k]['roledisplay'] = rolesarr.join(", ");

            }
        }
        return users;
    }
    onAfterSaveCell(row, cellName, cellValue) {
        //alert(`Save cell ${cellName} with value ${cellValue}`);

        let rowStr = '';
        for (const prop in row) {
            rowStr += prop + ': ' + row[prop] + '\n';
        }
        this.editUser(row.id, cellName, cellValue);
        //alert('Thw whole row :\n' + rowStr);
    }
    onRowSelect(row, isSelected, e) {
        let selected = new Selected();
        selected.rows = [row];
        selected.selected = isSelected;
        this.props.onRowSelect(selected);
    }
    onSelectAll(isSelected, rows) {
        let selected = new Selected();
        selected.rows = rows;
        selected.selected = isSelected;
        this.props.onRowSelect(selected);
    }
    emailFormatter(cell, row) {
        return cell;
        // return (<div>
        //     <Suggestion
        //         id={Math.random()}
        //         type="mitigation"
        //         value="abc"
        //         suggestions={[{name: "a", value: "a"},{name: "aa", value: "aa"},{name: "aaa", value: "aaa"},{name: "aaaa", value: "aaaa"},{name: "aaaaa", value: "aaaaa"}]} />
        // </div>)
    }
    render() {
        //console.log(this.props);
        let { users } = this.props;
        users = this.format(users);
        const roleTypes = [];
        for (let role of Constants.ROLES) {
            roleTypes.push(role.display);
        }
        const selectRowProp = {
            mode: 'checkbox',
            clickToSelect: true,
            unselectable: [1],
            bgColor: '#fff2fc',
            // onSelect: this.onRowSelect.bind(this),
            // onSelectAll: this.onSelectAll.bind(this)
        };
        const cellEditProp = {
            mode: 'click',
            blurToSave: true,
            afterSaveCell: this.onAfterSaveCell.bind(this)
        };
        return (
            <div>
                <div className="header-container">
                  <PageHeading title="User Management" url="/admin" />
                </div>
                <section className="paddedRow">
                    {/*cellEdit={ cellEditProp } selectRow={ selectRowProp }*/}
                    <BootstrapTable ref='userstable' data={ users } selectRow={ selectRowProp } pagination striped={ true } search={ true }>
                        <TableHeaderColumn dataField='id' isKey hidden dataSort={ true }>ID</TableHeaderColumn>
                        <TableHeaderColumn dataField='email' dataSort={ true } dataFormat={this.emailFormatter.bind(this)}>Email</TableHeaderColumn>
                        <TableHeaderColumn dataField='roledisplay' dataSort={ true } editable={ { type: 'select', options: { values: roleTypes } } }>Role</TableHeaderColumn>
                    </BootstrapTable>
                </section>
            </div>
        )
    }
}

class Selected{
    get rows() {
        return this._rows;
    }

    set rows(value) {
        this._rows = value;
    }

    get selected() {
        return this._selected;
    }

    set selected(value) {
        this._selected = value;
    }
    constructor(){
        this._rows = new Array();
        this._selected = new Boolean();
    }

}

const mapStateToProps = (state) => {
    return {
        users: state.adminReducer.users
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        getAllUsers: () => dispatch(getAllUsers())
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Authorization(AdminComponent, ['ADMINISTRATOR']));
