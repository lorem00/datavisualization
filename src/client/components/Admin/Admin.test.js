import React from 'react';
import { shallow } from 'enzyme';
import ShallowRenderer from 'react-test-renderer/shallow';

import { AdminComponent } from './Admin';

const users = [
    {
        id: '59dd13e1a6796b873e23acfd',
        account: '3a838c14-7a8e-4bc0-8e30-ea2372f6daec',
        email: 'ajalbani@wootcloud.com',
        roles: [
            'admin',
            'customer',
        ],
    }];

// Snapshot for Admin React Component
describe('>>>A D M I N --- Snapshot',()=>{
    it('+++capturing Snapshot of Home', () => {
        const renderer = new ShallowRenderer();
        const renderedValue = renderer.render(<AdminComponent users={users}/>);
        expect(renderedValue).toMatchSnapshot();
    });
});
// ******************************************************************************
describe('>>>A D M I N --- REACT-REDUX (Shallow + passing {users} directly)',()=>{
    let wrapper;
    let data;
    console.log('********* RUN npm test -- -u ***************');
    beforeEach(()=>{
        wrapper = shallow(<AdminComponent users={users} />);
        // TODO: shallow render table component;
        // let bootstrap = shallow(wrapper.find('BootstrapTable')[0]);
        data = wrapper.find('BootstrapTable').prop('data');
    });
    it('+++ expect the component to have 1 Bootstrap table', () => {
        expect(wrapper.find('BootstrapTable')).toHaveLength(1);
    });
    it('+++ expect the component to have matching data passed in from props', () => {
        expect(data).toHaveLength(users.length);
    });
    it('+++ expect the component first props account to match first rows account', () => {
        expect(data[0].account).toEqual(users[0].account);
    });
});
