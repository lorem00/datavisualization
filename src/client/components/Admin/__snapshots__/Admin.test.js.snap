// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>A D M I N --- Snapshot +++capturing Snapshot of Home 1`] = `
<div>
  <div
    className="header-container"
  >
    <PageHeading
      title="User Management"
      url="/admin"
    />
  </div>
  <section
    className="paddedRow"
  >
    <BootstrapTable
      autoCollapse={
        Object {
          "filter": false,
          "search": false,
          "sort": false,
        }
      }
      bodyContainerClass={null}
      bodyStyle={undefined}
      bordered={true}
      cellEdit={
        Object {
          "afterSaveCell": undefined,
          "beforeSaveCell": undefined,
          "blurToEscape": false,
          "blurToSave": false,
          "mode": "none",
          "nonEditableRows": undefined,
        }
      }
      columnFilter={false}
      condensed={false}
      containerClass={null}
      containerStyle={undefined}
      csvFileName="spreadsheet.csv"
      data={
        Array [
          Object {
            "account": "3a838c14-7a8e-4bc0-8e30-ea2372f6daec",
            "email": "ajalbani@wootcloud.com",
            "id": "59dd13e1a6796b873e23acfd",
            "roledisplay": "Administrator, Customer",
            "roles": Array [
              "admin",
              "customer",
            ],
          },
        ]
      }
      deleteRow={false}
      expandColumnOptions={
        Object {
          "expandColumnBeforeSelectColumn": true,
          "expandColumnComponent": undefined,
          "expandColumnVisible": false,
        }
      }
      expandComponent={undefined}
      expandableRow={undefined}
      exportCSV={false}
      fetchInfo={
        Object {
          "dataTotalSize": 0,
        }
      }
      footer={false}
      headerContainerClass={null}
      headerStyle={undefined}
      height="100%"
      hover={false}
      ignoreSinglePage={false}
      insertRow={false}
      keyBoardNav={false}
      maxHeight={undefined}
      multiColumnSearch={false}
      multiColumnSort={1}
      options={
        Object {
          "afterColumnFilter": undefined,
          "afterDeleteRow": undefined,
          "afterInsertRow": undefined,
          "afterSearch": undefined,
          "afterTableComplete": undefined,
          "alwaysShowAllBtns": false,
          "beforeShowError": undefined,
          "btnGroup": undefined,
          "clearSearch": false,
          "clearSearchBtn": undefined,
          "closeText": "Close",
          "defaultSearch": "",
          "defaultSortName": undefined,
          "defaultSortOrder": undefined,
          "deleteBtn": undefined,
          "deleteText": "Delete",
          "excludeCSVHeader": false,
          "expandBodyClass": null,
          "expandBy": "row",
          "expandParentClass": null,
          "expandRowBgColor": undefined,
          "expanding": Array [],
          "exportCSVBtn": undefined,
          "exportCSVSeparator": ",",
          "exportCSVText": "Export to CSV",
          "firstPage": "<<",
          "firstPageTitle": "first page",
          "handleConfirmDeleteRow": undefined,
          "hidePageListOnlyOnePage": false,
          "hideSizePerPage": false,
          "ignoreEditable": false,
          "insertBtn": undefined,
          "insertFailIndicator": "Validation errors, please check!",
          "insertModal": undefined,
          "insertModalBody": undefined,
          "insertModalFooter": undefined,
          "insertModalHeader": undefined,
          "insertText": "New",
          "keepSizePerPageState": false,
          "lastPage": ">>",
          "lastPageTitle": "last page",
          "nextPage": ">",
          "nextPageTitle": "next page",
          "noAutoBOM": true,
          "noDataText": undefined,
          "onExpand": undefined,
          "onMouseEnter": undefined,
          "onMouseLeave": undefined,
          "onRowClick": undefined,
          "onRowDoubleClick": undefined,
          "onRowMouseOut": undefined,
          "onRowMouseOver": undefined,
          "onSizePerPageList": undefined,
          "onlyOneExpanding": false,
          "page": undefined,
          "pageStartIndex": 1,
          "paginationPanel": undefined,
          "paginationPosition": "bottom",
          "paginationShowsTotal": false,
          "paginationSize": 5,
          "prePage": "<",
          "prePageTitle": "previous page",
          "printToolBar": true,
          "saveText": "Save",
          "searchDelayTime": undefined,
          "searchField": undefined,
          "searchPanel": undefined,
          "searchPosition": "right",
          "showSelectedOnlyBtn": undefined,
          "sizePerPage": undefined,
          "sizePerPageDropDown": undefined,
          "sizePerPageList": Array [
            10,
            25,
            30,
            50,
          ],
          "sortIndicator": true,
          "sortName": undefined,
          "sortOrder": undefined,
          "toolBar": undefined,
          "toolbarPosition": "top",
          "withFirstAndLast": true,
          "withoutNoDataText": false,
        }
      }
      pagination={true}
      printable={false}
      replace={false}
      scrollTop={undefined}
      search={true}
      searchPlaceholder={undefined}
      selectRow={
        Object {
          "bgColor": "#fff2fc",
          "clickToSelect": true,
          "mode": "checkbox",
          "unselectable": Array [
            1,
          ],
        }
      }
      strictSearch={undefined}
      striped={true}
      tableBodyClass={null}
      tableContainerClass={null}
      tableFooterClass={null}
      tableHeaderClass={null}
      tableStyle={undefined}
      trClassName=""
      trStyle={undefined}
      version="3"
      withoutTabIndex={false}
    >
      <TableHeaderColumn
        className=""
        columnClassName=""
        columnTitle={false}
        csvFieldType="string"
        csvFormat={undefined}
        csvHeader={undefined}
        dataAlign="left"
        dataField="id"
        dataFormat={undefined}
        dataSort={true}
        defaultASC={false}
        editColumnClassName=""
        editTdAttr={undefined}
        editable={true}
        expandable={true}
        filter={undefined}
        filterFormatted={false}
        filterValue={undefined}
        formatExtraData={undefined}
        headerAlign={undefined}
        headerTitle={true}
        hidden={true}
        hiddenOnInsert={false}
        invalidEditColumnClassName=""
        isKey={true}
        keyValidator={false}
        onSort={undefined}
        searchable={true}
        sort={undefined}
        sortFunc={undefined}
        sortFuncExtraData={undefined}
        sortIndicator={true}
        tdAttr={undefined}
        tdStyle={undefined}
        thStyle={undefined}
        width={null}
      >
        ID
      </TableHeaderColumn>
      <TableHeaderColumn
        className=""
        columnClassName=""
        columnTitle={false}
        csvFieldType="string"
        csvFormat={undefined}
        csvHeader={undefined}
        dataAlign="left"
        dataField="email"
        dataFormat={[Function]}
        dataSort={true}
        defaultASC={false}
        editColumnClassName=""
        editTdAttr={undefined}
        editable={true}
        expandable={true}
        filter={undefined}
        filterFormatted={false}
        filterValue={undefined}
        formatExtraData={undefined}
        headerAlign={undefined}
        headerTitle={true}
        hidden={false}
        hiddenOnInsert={false}
        invalidEditColumnClassName=""
        isKey={false}
        keyValidator={false}
        onSort={undefined}
        searchable={true}
        sort={undefined}
        sortFunc={undefined}
        sortFuncExtraData={undefined}
        sortIndicator={true}
        tdAttr={undefined}
        tdStyle={undefined}
        thStyle={undefined}
        width={null}
      >
        Email
      </TableHeaderColumn>
      <TableHeaderColumn
        className=""
        columnClassName=""
        columnTitle={false}
        csvFieldType="string"
        csvFormat={undefined}
        csvHeader={undefined}
        dataAlign="left"
        dataField="roledisplay"
        dataFormat={undefined}
        dataSort={true}
        defaultASC={false}
        editColumnClassName=""
        editTdAttr={undefined}
        editable={
          Object {
            "options": Object {
              "values": Array [
                "Administrator",
                "Customer",
              ],
            },
            "type": "select",
          }
        }
        expandable={true}
        filter={undefined}
        filterFormatted={false}
        filterValue={undefined}
        formatExtraData={undefined}
        headerAlign={undefined}
        headerTitle={true}
        hidden={false}
        hiddenOnInsert={false}
        invalidEditColumnClassName=""
        isKey={false}
        keyValidator={false}
        onSort={undefined}
        searchable={true}
        sort={undefined}
        sortFunc={undefined}
        sortFuncExtraData={undefined}
        sortIndicator={true}
        tdAttr={undefined}
        tdStyle={undefined}
        thStyle={undefined}
        width={null}
      >
        Role
      </TableHeaderColumn>
    </BootstrapTable>
  </section>
</div>
`;
